### PR TITLE
sml clang 12 patch

### PR DESCRIPTION
**Pull request by werto87**
Specify library name and version:  **sml/1.1.4**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

---
sml clang 12 support

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by prince-chrismc** [comment]
Not all of your commits have the same authorship, the CLA is not passing https://github.com/conan-io/conan-center-index/pull/5652#issuecomment-849715763

File: recipes/sml/all/conanfile.py [hunk 1]
```diff
@@ -13,37 +13,43 @@ class SMLConan(ConanFile):
...
     def package(self):
-        self.copy(pattern="*", dst="include", src=os.path.join(self._source_subfolder, "include"))
+        self.copy(pattern="*", dst="include",
+                  src=os.path.join(self._source_subfolder, "include"))
         self.copy("*LICENSE.md", dst="licenses", keep_path=False)
 
-    def package_id(self):
-        self.info.header_only()
+        def package_id(self):
+            self.info.header_only()
```
Comment (prince-chrismc) [inline recipes/sml/all/conanfile.py:55]: Please revert this

File: recipes/sml/all/conanfile.py [hunk 2]
```diff
@@ -13,37 +13,43 @@ class SMLConan(ConanFile):
...
+        if not minimum_version:
+            self.output.warn(
+                "SML requires C++14. Your compiler is unknown. Assuming it supports C++14.")
+        elif tools.Version(self.settings.compiler.version) < minimum_version:
+            raise ConanInvalidConfiguration(
+                "SML requires C++14, which your compiler does not support.")
 
     def source(self):
         tools.get(**self.conan_data["sources"][self.version])
+        tools.patch(patch_file="patches/clang12.patch")
```
Comment (prince-chrismc) [inline recipes/sml/all/conanfile.py:45]: This needs to be moved to build like 

https://github.com/conan-io/conan-center-index/blob/41d4fbb8c218f9d938ccb8b9d888bff52002e62e/recipes/libmorton/all/conanfile.py#L27-L29

File: recipes/sml/all/conanfile.py [hunk 3]
```diff
@@ -13,37 +13,43 @@ class SMLConan(ConanFile):
     url = "https://github.com/conan-io/conan-center-index"
     settings = "compiler"
     no_copy_source = True
```
Comment (prince-chrismc) [inline recipes/sml/all/conanfile.py:15]: To have a build method =)

```suggestion
```

File: recipes/sml/all/conandata.yml [hunk 4]
```diff
@@ -2,3 +2,7 @@ sources:
   "1.1.4":
     url: https://github.com/boost-ext/sml/archive/v1.1.4.tar.gz
     sha256: 897c78077f5a4d22b0de253b8689a8ea6d8adfba8d7b4877d6f5c76277e00976
+patches:
+  "1.1.4":
+    - patch_file: "patches/clang12.patch"
```
Comment (prince-chrismc) [inline recipes/sml/all/conandata.yml:7]: We typically like to make patches more descriptive... like "0001-fix-clang12-error.patch" just as a suggestion

Also has this been submitted to upstream?

File: recipes/sml/all/conanfile.py [hunk 5]
```diff
@@ -13,37 +13,43 @@ class SMLConan(ConanFile):
...
+                "SML requires C++14. Your compiler is unknown. Assuming it supports C++14.")
+        elif tools.Version(self.settings.compiler.version) < minimum_version:
+            raise ConanInvalidConfiguration(
+                "SML requires C++14, which your compiler does not support.")
 
     def source(self):
         tools.get(**self.conan_data["sources"][self.version])
+        tools.patch(patch_file="patches/clang12.patch")
         extracted_dir = "sml-" + self.version
         os.rename(extracted_dir, self._source_subfolder)
```
Comment (prince-chrismc) [inline recipes/sml/all/conanfile.py:47]: Completely unrelated but this is the new convention

```suggestion
        tools.get(**self.conan_data["sources"][self.version], strip_root=True, destination=self._source_subfolder)
```

**Review by prince-chrismc — CHANGES_REQUESTED** [review]
Thanks for the contribution... Just a couple small technical details to iron out

**Comment by werto87** [comment]
> Not all of your commits have the same authorship, the CLA is not passing [#5652 (comment)](https://github.com/conan-io/conan-center-index/pull/5652#issuecomment-849715763)

Hi prince-chrismc,

I do not know how to fix this isue.
I can not see a question mark here
"To read a message about why the commit is not linked, hover over the blue to the right of the username."

**Comment by prince-chrismc** [comment]
You need to change the author of all the commits to match the GitHub user (easy way)

Found this how to
https://stackoverflow.com/questions/750172/how-to-change-the-author-and-committer-name-and-e-mail-of-multiple-commits-in-gi

**Comment by prince-chrismc** [comment]
I still see bad authored commits in the UI

**Comment by werto87** [comment]
> You need to change the author of all the commits to match the GitHub user (easy way)
> 
> Found this how to
> https://stackoverflow.com/questions/750172/how-to-change-the-author-and-committer-name-and-e-mail-of-multiple-commits-in-gi



> I still see bad authored commits in the UI

I could work on your comments. After that I could cancel everything and make a new request

**Comment by prince-chrismc** [comment]
Just squash and force push, it's the easiest way... a new PR will have the same problem

**Comment by werto87** [comment]
> Just squash and force push, it's the easiest way... a new PR will have the same problem

I tried this but it seems not to work.

**Comment by prince-chrismc** [comment]
> > Just squash and force push, it's the easiest way... a new PR will have the same problem
> 
> I tried this but it seems not to work.

Ahhh I see you made a small mistake somewhere. Let's see if I can help you with the git commands 😸 

⚠️ I did not test this locally but it's from my script... double check

```sh
git reset --mixed f2542cd56c3483a604fe32b68022f6ab26996eef
git commit -a -m "squashed"
# 👇 Double check all your changes are still present
git push -f
```

**Comment by prince-chrismc** [comment]
CLA is passing !

File: recipes/sml/all/conanfile.py (hunk 2 above)
Comment (werto87) [reply recipes/sml/all/conanfile.py:45]: I do not know where to put the for patches. There is no build function. How to deal with patches if its header only?

Comment (werto87) [reply recipes/sml/all/conanfile.py:45]: Is the idea to change the source in the build step?

File: recipes/sml/all/conandata.yml (hunk 4 above)
Comment (werto87) [reply recipes/sml/all/conandata.yml:7]: > Also has this been submitted to upstream

There is an issue on the project github. "https://github.com/boost-ext/sml/issues/447"

**Comment by prince-chrismc** [comment]
You can follow this example

https://github.com/conan-io/conan-center-index/blob/41d4fbb8c218f9d938ccb8b9d888bff52002e62e/recipes/libmorton/all/conanfile.py#L27

File: recipes/sml/all/conanfile.py [hunk 6]
```diff
@@ -12,37 +12,45 @@ class SMLConan(ConanFile):
...
+            self.output.warn(
+                "SML requires C++14. Your compiler is unknown. Assuming it supports C++14.")
+        elif tools.Version(self.settings.compiler.version) < minimum_version:
+            raise ConanInvalidConfiguration(
+                "SML requires C++14, which your compiler does not support.")
+
+    def build(self):
+        for patch in self.conan_data.get("patches", {}).get(self.version, []):
+            tools.patch(**patch)
+        # tools.patch(patch_file="patches/0001-fix-clang12-error.patch")
```
Comment (prince-chrismc) [inline recipes/sml/all/conanfile.py:49]:
```suggestion
```

**Review by prince-chrismc — APPROVED** [review]
Extra comment but that's it

**Review by SSE4 — APPROVED** [review]

**Review by jgsogo — APPROVED** [review]

**Pull request merged**